lib/storage/data/file/store.js

```diff
@@ -0,0 +1,274 @@
+     * @param {Object} logConfig - logging configuration
+     */
+    constructor(dataConfig, logConfig) {
+        this.logger = new Logger('DataFileStore', logConfig);
+        this.dataPath = dataConfig.dataPath;
+        this.noSync = dataConfig.noSync || false;
+    }
+
+    /**
+     * Setup the storage backend before starting to read or write
+     * files in it.
+     *
+     * The function ensures that dataPath is accessible and
+     * pre-creates the directory hashes under dataPath.
+     *
+     * @param {function} callback - called when done with no argument
+     * @return {undefined}
+     */
+    setup(callback) {
+        fs.accessSync(this.dataPath, fs.F_OK | fs.R_OK | fs.W_OK);
+
+        // Create FOLDER_HASH subdirectories
+        const subDirs = Array.from({ length: FOLDER_HASH },
+                                   (v, k) => (k).toString());
+        this.logger.info(`pre-creating ${subDirs.length} subdirs...`);
+        if (!this.noSync) {
+            storageUtils.setDirSyncFlag(this.dataPath, this.logger);
+        }
+        async.eachSeries(subDirs, (subDirName, next) => {
+            fs.mkdir(`${this.dataPath}/${subDirName}`, err => {
+                // If already exists, move on
+                if (err && err.code !== 'EEXIST') {
+                    return next(err);
+                }
+                return next();
+            });
+        },
+        err => {
+            assert.strictEqual(err, null, `Error creating data files ${err}`);
+            this.logger.info('data file store init complete, ' +
+                             'go forth and store data.');
+            callback();
+        });
+    }
+
+    /**
+     * Get the filesystem path to a stored object file from its key
+     *
+     * @param {String} key - the object key
+     * @return {String} the absolute path to the file containing the
+     *   object contents
+     */
+    getFilePath(key) {
+        const hash = stringHash(key);
+        const folderHashPath = ((hash % FOLDER_HASH)).toString();
+        return `${this.dataPath}/${folderHashPath}/${key}`;
+    }
+
+    /**
+     * Put a new object to the storage backend
+     *
+     * @param {stream.Readable} dataStream - input stream providing the
+     *   object data
+     * @param {Number} size - Total byte size of the data to put
+     * @param {Logger} log - logging object
+     * @param {DataFileStore~putCallback} callback - called when done
+     * @return {undefined}
+     */
+    put(dataStream, size, log, callback) {
+        const key = crypto.pseudoRandomBytes(20).toString('hex');
+        const filePath = this.getFilePath(key);
+        log.debug('starting to write data', { key, filePath });
+        dataStream.pause();
+        fs.open(filePath, 'wx', (err, fd) => {
+            if (err) {
+                log.error('error opening filePath', { error: err });
+                return callback(err);
+            }
+            const cbOnce = jsutil.once(callback);
```

Comment:
We have to guard against calling multiple times here, because we're not controlling how errors events are generated, both from `dataStream` and from `fileStream`. Both may actually generate errors until they are destroyed, and multiple errors are also possible AFAIK.